internal/service/db/impl.go

```diff
@@ -0,0 +1,268 @@
+) ([]*upstreamv0.ServerJSON, error) {
+	// TODO: implement
+	options := &service.ListServersOptions{}
+	for _, opt := range opts {
+		if err := opt(options); err != nil {
+			return nil, err
+		}
+	}
+
+	decoded, err := base64.StdEncoding.DecodeString(options.Cursor)
+	if err != nil {
+		return nil, err
+	}
+	nextTime, err := time.Parse(time.RFC3339, string(decoded))
+	if err != nil {
+		return nil, err
+	}
+
+	querierFunc := func(querier sqlc.Querier) ([]helper, error) {
+		servers, err := querier.ListServers(
+			ctx,
+			sqlc.ListServersParams{
+				Next: &nextTime,
+				Size: int64(options.Limit),
+			},
+		)
+
+		if err != nil {
+			return nil, err
+		}
+
+		helpers := make([]helper, len(servers))
+		for i, server := range servers {
+			helpers[i] = listServersRowToHelper(server)
+		}
+
+		return helpers, nil
+	}
+
+	return s.sharedListServers(ctx, querierFunc)
+}
+
+// ListServerVersions implements RegistryService.ListServerVersions
+func (s *dbService) ListServerVersions(
+	ctx context.Context,
+	opts ...service.Option[service.ListServerVersionsOptions],
+) ([]*upstreamv0.ServerJSON, error) {
+	options := &service.ListServerVersionsOptions{}
+	for _, opt := range opts {
+		if err := opt(options); err != nil {
+			return nil, err
+		}
+	}
+
+	querierFunc := func(querier sqlc.Querier) ([]helper, error) {
+		servers, err := querier.ListServerVersions(
+			ctx,
+			sqlc.ListServerVersionsParams{
+				Name: options.Name,
+				Next: options.Next,
+				Prev: options.Prev,
+				Size: int64(options.Limit),
+			},
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		helpers := make([]helper, len(servers))
+		for i, server := range servers {
+			helpers[i] = listServerVersionsRowToHelper(server)
+		}
+
+		return helpers, nil
+	}
+
+	return s.sharedListServers(ctx, querierFunc)
+}
+
+// GetServer returns a specific server by name
+func (s *dbService) GetServerVersion(
+	ctx context.Context,
+	opts ...service.Option[service.GetServerVersionOptions],
+) (*upstreamv0.ServerJSON, error) {
+	options := &service.GetServerVersionOptions{}
+	for _, opt := range opts {
+		if err := opt(options); err != nil {
+			return nil, err
+		}
+	}
+
+	querierFunc := func(querier sqlc.Querier) ([]helper, error) {
+		server, err := querier.GetServerVersion(
+			ctx,
+			sqlc.GetServerVersionParams{
+				Name:    options.Name,
+				Version: options.Version,
+			},
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		return []helper{getServerVersionRowToHelper(server)}, nil
+	}
+
+	res, err := s.sharedListServers(ctx, querierFunc)
+	if err != nil {
+		return nil, err
+	}
+
+	return res[0], nil
+}
+
+func (s *dbService) sharedListServers(
+	ctx context.Context,
+	querierFunc func(querier sqlc.Querier) ([]helper, error),
+) ([]*upstreamv0.ServerJSON, error) {
+	tx, err := s.pool.BeginTx(ctx, pgx.TxOptions{
+		IsoLevel:   pgx.ReadCommitted,
+		AccessMode: pgx.ReadOnly,
+	})
+	if err != nil {
+		return nil, err
+	}
+	defer tx.Rollback(ctx)
```

Comment:
nit: Would this ignore rollback errors? Something for future though, thats a can of worms in and of itself

```suggestion
      defer func() {
          if rollbackErr := tx.Rollback(ctx); rollbackErr != nil &&
              !errors.Is(rollbackErr, pgx.ErrTxClosed) {
              // Log the rollback error (add proper logging)
              _ = rollbackErr // TODO: Add logging
          }
      }()
````